core/trino-main/src/main/java/io/trino/operator/join/unspilled/JoinProbe.java

```diff
@@ -108,23 +100,81 @@ public Page getPage()
-            if (probePage.getBlock(i).isNull(position)) {
-                return true;
+        List<Block> nullableBlocks = IntStream.range(0, probePage.getChannelCount())
+                .mapToObj(i -> probePage.getBlock(i))
+                .filter(Block::mayHaveNull)
+                .collect(toImmutableList());
+
+        long[] joinPositionCache = new long[positionCount];
+        if (!nullableBlocks.isEmpty()) {
+            Arrays.fill(joinPositionCache, -1);
+            boolean[] isNull = new boolean[positionCount];
+            int nonNullCount = getIsNull(nullableBlocks, positionCount, probePage, isNull);
+            if (nonNullCount < positionCount) {
+                // We only store positions that are not null
+                int[] positions = new int[nonNullCount];
+                nonNullCount = 0;
+                for (int i = 0; i < positionCount; i++) {
+                    if (!isNull[i]) {
+                        positions[nonNullCount] = i;
+                    }
+                    // This way less code is in the if branch and CPU should be able to optimize branch prediction better
+                    nonNullCount += isNull[i] ? 0 : 1;
+                }
+                if (probeHashBlock != null) {
+                    long[] hashes = new long[positionCount];
+                    for (int i = 0; i < positionCount; i++) {
+                        hashes[i] = BIGINT.getLong(probeHashBlock, i);
+                    }
+                    lookupSource.getJoinPosition(positions, probePage, page, hashes, joinPositionCache);
+                }
+                else {
+                    lookupSource.getJoinPosition(positions, probePage, page, joinPositionCache);
+                }
+                return joinPositionCache;
+            } // else fall back to non-null path
+        }
+        int[] positions = new int[positionCount];
+        for (int i = 0; i < positionCount; i++) {
+            positions[i] = i;
+        }
+        if (probeHashBlock != null) {
+            long[] hashes = new long[positionCount];
+            for (int i = 0; i < positionCount; i++) {
+                hashes[i] = BIGINT.getLong(probeHashBlock, i);
             }
+            lookupSource.getJoinPosition(positions, probePage, page, hashes, joinPositionCache);
+        }
+        else {
+            lookupSource.getJoinPosition(positions, probePage, page, joinPositionCache);
         }
-        return false;
+
+        return joinPositionCache;
     }
 
-    private static boolean probeMayHaveNull(Page probePage)
+    private static int getIsNull(List<Block> nullableBlocks, int positionCount, Page probePage, boolean[] isNull)
     {
-        for (int i = 0; i < probePage.getChannelCount(); i++) {
-            if (probePage.getBlock(i).mayHaveNull()) {
-                return true;
+        for (int i = 0; i < nullableBlocks.size() - 1; i++) {
+            Block block = probePage.getBlock(i);
```

Comment:
This looks suspicious, should this be `Block block = nullableBlocks.get(i)` ?
Would it be simpler to pass in list of indexes of nullable blocks of the probe page rather than the list of nullable blocks ?

Reply:
> This looks suspicious, should this be Block block = nullableBlocks.get(i) ?
You are right. Fixed
> Would it be simpler to pass in list of indexes of nullable blocks of the probe page rather than the list of nullable blocks ?
I prefer my way, because there is no two-level indexing.

Reply:
ping